python/social_bot/envs/grocery_ground.py

```diff
@@ -63,22 +65,167 @@ def __init__(self, reward_shaping=False, random_goal=False, **kwargs):
             'coke_can', 'table', 'bookshelf', 'car_wheel',
             'plastic_cup', 'beer', 'hammer'
         ]
+        self._pos_list = list(itertools.product(range(-5, 5), range(-5, 5)))
+        self._pos_list.remove((0, 0))
         self.task_vocab = self.task_vocab + self._objects_in_world + self._objects_to_insert
 
-    def run(self, agent, world):
+    def setup(self, world, agent_name):
+        """
+        Setting things up during the initialization
+        """
+        self._world = world
+        self._agent = self._world.get_agent()
+        self._agent_name = agent_name
+        self._insert_objects(self._objects_to_insert)
+
+    def reset(self):
+        """
+        Reset each time after environment is reseted
+        """
+        self._random_move_objects()
         if self._random_goal:
             random_id = random.randrange(len(self._objects_to_insert))
             self.set_goal_name(self._objects_to_insert[random_id])
-        yield from super().run(agent, world)
 
-    def get_object_list(self):
+    def _insert_objects(self, object_list):
+        obj_num = len(object_list)
+        for obj_id in range(obj_num):
+            model_name = object_list[obj_id]
+            self._world.insertModelFile('model://' + model_name)
+            logging.debug('model ' + model_name + ' inserted')
+            self._world.step(20)
+            # Sleep for a while waiting for Gazebo server to finish the inserting
+            # operation. Or the model may not be completely inserted, boost will
+            # throw 'px!=0' error when set_pose/get_pose of the model is called
+            time.sleep(0.2)
+
+    def _random_move_objects(self, random_range=10.0):
+        obj_num = len(self._objects_to_insert)
+        obj_pos_list = random.sample(self._pos_list, obj_num)
+        for obj_id in range(obj_num):
+            model_name = self._objects_to_insert[obj_id]
+            loc = (obj_pos_list[obj_id][0], obj_pos_list[obj_id][1], 0)
+            pose = (np.array(loc), (0, 0, 0))
+            self._world.get_model(model_name).set_pose(pose)
+
+    def task_specific_observation(self):
```

Comment:
since it's part of Task interface, we should also add it and document it in Task class.

Reply:
Let task class be simple and more general purpose. Some task may not need functions like task-specific observations. We can rename GoalTask to GoalBaseTask, let all other goal related tasks use this base task, and document the extra interface in GoalBaseTask.